api/service/projects_service.go

```diff
@@ -250,3 +279,122 @@ func (service *projectsService) filterAuthorizedProjects(ctx context.Context, pr
+	if service.updateProjectConfig.Endpoint == "" {
+		return project, nil, nil
+	}
+
+	if service.updateProjectConfig.PayloadTemplate == "" || service.updateProjectConfig.ResponseTemplate == "" {
+		return project, nil, nil
+	}
+
+	payload, err := generateRequestPayload(project, service.updateProjectConfig.PayloadTemplate)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error generating request payload: %w", err)
+	}
+
+	resp, err := sendUpdateRequest(service.updateProjectConfig.Endpoint, payload)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error sending update request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	response, err := processResponseTemplate(resp, service.updateProjectConfig.ResponseTemplate)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error processing response template: %w", err)
+	}
+
+	return project, response, nil
+}
+
+func generateRequestPayload(project *models.Project, templateString string) (map[string]interface{}, error) {
+	tmpl, err := template.New("requestPayload").Parse(templateString)
+	if err != nil {
+		return nil, err
+	}
+	var payload bytes.Buffer
+	if err := tmpl.Execute(&payload, project); err != nil {
+		return nil, err
+	}
+	var result map[string]interface{}
+	if err := json.Unmarshal(payload.Bytes(), &result); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func sendUpdateRequest(url string, payload map[string]interface{}) (*http.Response, error) {
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(payloadBytes))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+func processResponseTemplate(response *http.Response, templateString string) (map[string]interface{}, error) {
+	var data map[string]interface{}
+	if err := json.NewDecoder(response.Body).Decode(&data); err != nil {
+		return nil, err
+	}
+
+	tmpl, err := template.New("responsePayload").Parse(templateString)
+	if err != nil {
+		return nil, err
+	}
+
+	var responseText bytes.Buffer
+	if err := tmpl.Execute(&responseText, data); err != nil {
+		return nil, err
+	}
+
+	var result map[string]interface{}
+	if err := json.Unmarshal(responseText.Bytes(), &result); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// areBlacklistedLabelsChanged check if any key in labels is blacklisted
+func (service *projectsService) areBlacklistedLabelsChanged(projectID models.ID, newLabels []models.Label,
+	blacklist map[string]bool) (bool, error) {
+	existingProject, err := service.FindByID(projectID)
+	if err != nil {
+		return false, fmt.Errorf("error fetching project with id %s: %w", projectID, err)
+	}
+
+	for _, existingLabel := range existingProject.Labels {
+		if blacklist[existingLabel.Key] {
+			found := false
+			for _, newLabel := range newLabels {
```

Comment:
I would think creating a hash map with `newLabels` would help with the look up speed. Imagine having n old labels, and all n are blacklisted, and n new labels, though none been changed. This will take O(n^2) time to complete. Creating a hash map of `newLabels` would only take O(n) time, but would give you O(1) lookups when looping through the old labels, giving you at worst O(n).